File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int longest = 0;
        int startWindow = 0;
        Set<Character> seen = new HashSet<>();
        for(int endWindow = 0;endWindow<s.length();endWindow++){
            while(seen.contains(s.charAt(endWindow))){
                seen.remove(s.charAt(startWindow));
                startWindow++;
            }
            seen.add(s.charAt(endWindow));
            longest = Math.max(longest, endWindow - startWindow + 1);
        }
        return longest;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        sets = set()
        left = longest = 0
        for right in range(0,len(s)):
            while s[right] in sets:
                sets.remove(s[left])
                left+=1
            sets.add(s[right])
            longest = max(longest, len(sets))
        return longest
